### PR TITLE
Collection of Noah-MP bug fixes from WRFv4.1

### PR DESCRIPTION
**Pull request by barlage**
TYPE: bug fix

KEYWORDS: Noah-MP

SOURCE: Michael Barlage (NCAR)

DESCRIPTION OF CHANGES:

Add Noah-MP bug fixes to unify community code with WRF v4.1.

When run with LSM_only, max differences in 7-day run using Croton domain:
accumulated SFCRNOFF: 0.2 mm
accumulated UGDRNOFF: 0.02 mm

When run with LSM_only, max differences in 7-day run using Taylor Park domain:
accumulated SFCRNOFF: 0.22 mm
accumulated UGDRNOFF: 0.003 mm

When run with hydro, max differences in 7-day run using Croton domain:
accumulated SFCRNOFF: 1.5 mm
accumulated UGDRNOFF: 0.25 mm
accumulated ET: 0.12 mm

When run with hydro, max differences in 7-day run using Taylor Park domain:
accumulated SFCRNOFF: 0.015 mm
accumulated UGDRNOFF: 0.001 mm
accumulated ET: 0.02 mm

This PR addresses 5. in #333 

LIST OF MODIFIED FILES:
M module_sf_noahmplsm.F
M module_sf_noahmp_glacier.F
M module_sf_noahmp_groundwater.F

TESTS CONDUCTED:

1. One-year simulation (Cheyenne, PGI) in two calibration basins

### CODE REVIEW

File: trunk/NDHMS/Land_models/NoahMP/phys/module_sf_noahmp_groundwater.F [hunk 1]
```diff
@@ -250,8 +250,7 @@ SUBROUTINE LATERALFLOW  (ISLTYP,WTD,QLAT,FDEPTH,TOPO,LANDMASK,DELTAT,AREA &
   DATA KLATFACTOR /2.,3.,4.,10.,10.,12.,14.,20.,24.,28.,40.,48.,2.,0.,10.,0.,20.,2.,2./
 
   REAL,    PARAMETER :: PI = 3.14159265 
-! REAL,    PARAMETER :: FANGLE = SQRT(TAN(PI/8.))/(2.*SQRT(2.))
-  REAL,    PARAMETER :: FANGLE = 0.45508986056   ! f95 does not permit real intrinsics in init expressions
+  REAL,    PARAMETER :: FANGLE = 0.22754493   ! = 0.5*sqrt(0.5*tan(pi/8))
```
Comment (jmccreight) [inline trunk/NDHMS/Land_models/NoahMP/phys/module_sf_noahmp_groundwater.F]: Really, a new fangle parameter?
😂

**Comment by jmccreight** [comment]
@barlage can you just summarize these? or maybe you have a summary from a different repo that you can paste here.
My comments are mostly just wondering what is changing/improving.

**Comment by jmccreight** [comment]
@barlage The diffs are not insignificant. Probably worth considering. I'm surprised you got no differences on a small domain.... croton test domain is perty small.

**Comment by barlage** [comment]
@jmccreight are there urban or glacier points in croton? if not, there are only three changes to module_sf_noahmplsm that are relevant to the NWM configuration:

new code lines 2359, 2360: and that only occurs if soil moisture becomes 0
new code lines 3553, 4009: only for diagnostic calculation of 2m output and shouldn't change results
new code line 4244: I have seen this change make some big localized differences in fluxes if exchange resistance gets low, for example when the wind speed gets very high (don't know if that happens in croton)

**Comment by kafitzgerald** [comment]
@barlage the Croton test domains do include urban pixels.  

The categories for the NWM domain are the following: 
1.0 2.0 7.0 11.0 16.0 18.0

and for the gridded domain:
1.0 2.0 3.0 11.0 16.0 18.0

The max wind speed in the forcings for the Croton domain is 20.3 m/s, so quite high.

**Comment by jmccreight** [comment]
@rcabell will run 48-hr conus test.

**Comment by rcabell** [comment]
Long-range CONUS testing only failed Regression (expected)

**Comment by barlage** [comment]
@rcabell @kafitzgerald can you hold off on merging this, I'm working on some limits on snow that I could add to this, unless you want me to put it in as a separate PR

**Comment by rcabell** [comment]
> @rcabell @kafitzgerald can you hold off on merging this, I'm working on some limits on snow that I could add to this, unless you want me to put it in as a separate PR

@barlage — I'm redoing some testing that failed originally due to (presumably) Cheyenne issues. I'll hold off until updates are in place.

**Comment by rcabell** [comment]
Attached are the results of the conus testing with the nwm_ana configuration. Regression results with `master` do not match, which is expected. @barlage, @jmccreight, @aubreyd: if the results of these tests seem reasonable based on known changes to the underlying physics, please comment with recommendation to merge. Thanks!

[PR319-diffs.txt](https://github.com/NCAR/wrf_hydro_nwm_public/files/3622917/PR319-diffs.txt)
[PR319_nwm-ana_CONUS.pdf](https://github.com/NCAR/wrf_hydro_nwm_public/files/3622919/PR319_nwm-ana_CONUS.pdf)

**Comment by kafitzgerald** [comment]
@rcabell Do we still have the output files from the NoahMP bugfix CONUS test?  If we do, I'll go ahead and create some maps of the diffs for reference.  The diffs aren't small and since the changes are in the LSM they propagate through most of the outputs.  That said, if @aubreyd and @barlage have already taken a look and are good with the diffs I'll happily skip that and check the box.

**Comment by rcabell** [comment]
The output is in: 

    /glade/u/home/rcabell/scratch2/testing/newconus-48-PR319_ana/nwm_ana

**Comment by aubreyd** [comment]
I have my earlier one-month CONUS test results for this patch, but I only looked at streamflow: https://docs.google.com/presentation/d/1KemlFg1uJbQPtnZw8DiW98tMCZG9nWQlWeF80GmgAfg/edit?usp=sharing

**Comment by kafitzgerald** [comment]
Here are some plots of the larger diffs from the 48 hour CONUS test on the land grid (mostly of energy balance variables): [Google Drive link](https://drive.google.com/drive/folders/1_CHpE-rJP5AONaVnYrbeGJa1vK301sQE?usp=sharing)

It looks like the diffs overall are smaller and more localized than I had expected based upon the table of stats, but consistent with what Aubrey saw in her longer test.  There were also some (locally) large differences in snow related variables though limited spatially and largely because of snow presence / absence.  I'm guessing there would be more of these earlier in the melt season.  That said, I believe these diffs were all expected.

**Pull request merged**